lib/spack/spack/solver/asp.py

```diff
@@ -2032,14 +2036,16 @@ def reorder_flags(self):
             sorted_sources = sorted(sources, key=lambda s: order.index(s))
 
             # add flags from each source, lowest to highest precedence
-            flags = collections.defaultdict(lambda: [])
+            flags = []
             for source_name in sorted_sources:
                 source = cmd_specs[source_name]
-                for name, flag_list in source.compiler_flags.items():
-                    extend_flag_list(flags[name], flag_list)
+                flags.extend(source.compiler_flags[flag_type])
 
-            check_same_flags(spec.compiler_flags, flags)
-            spec.compiler_flags.update(flags)
+            assert set(spec.compiler_flags[flag_type]) == set(flags), "%s does not equal %s" % (
+                spec.compiler_flags[flag_type],
+                flags,
+            )
+            spec.compiler_flags.update({flag_type: flags})
```

Comment:
This probably needs an update for when compiler flags also came from the compiler definition, but that can wait for a follow-on, it's a preexisting bug.